local server = require(game.ServerStorage.server)

local NetPackageService = server.Services.NetPackageService()
local NetOwnerService = server.Services.NetOwnerService()
local NetCoordinateService = server.Services.NetCoordinateService()
local NetInputService = server.Services.NetInputService()
local NetContextGatewayService = server.Services.NetContextGatewayService()

local GameMapService = server.Services.GameMapService()
local GameStunService = server.Services.GameStunService()
local GameCollisionService = server.Services.GameCollisionService()
local GameHitScanService = server.Services.GameHitScanService()
local GameContextBusService = server.Services.GameContextBusService()
local GameWorldsService = server.Services.GameWorldsService()

GameMapService.Service:SetActive(true)
GameStunService.Service:SetActive(true)
GameCollisionService.Service:SetActive(true)
GameHitScanService.Service:SetActive(true)
GameHitScanService.Service:SetLogActive(true)
GameContextBusService.Service:SetActive(true)
GameContextBusService.Service:SetLogActive(true)
GameWorldsService.Service:SetActive(true)
GameWorldsService.Service:SetLogActive(true)

NetPackageService.Service:SetActive(true)
NetPackageService.Service:SetLogActive()
NetOwnerService.Service:SetAuto(true)
NetOwnerService.Service:SetActive(true)
NetCoordinateService.Service:SetActive(true)
NetCoordinateService.Service:SetAuto(true)
NetInputService.Service:SetActive(true)
NetInputService.Service:SetAuto(true)
NetInputService.Service:SetLogActive(true)
NetContextGatewayService.Service:SetActive(true)
NetContextGatewayService.Service:SetLogActive(true)

local TestZone = GameMapService.Service:GetMap('TestZone')
TestZone:MapLoad() 

local world = GameWorldsService.Service:GetWorld('game.world')
local Fortuna = server.Characters.Fortuna()

world:DoForTarget(Fortuna:DoCreate(world), function(World) 
	Fortuna:DoAddComponents(World)
	Fortuna:DoAddAbilities(World)
	print(World:IsComponentsExists(Fortuna.Fabrics.model.Name))
    Fortuna.Fabrics.model:DoCatchFrom(World, function(Component) 
        print(Component)
    end)
end)


warn(world)